modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java

```diff
@@ -0,0 +1,327 @@
+// TODO: IGNITE-20637 Need integration with the IgniteImpl
+public class IndexAvailabilityController implements ManuallyCloseable {
+    private static final IgniteLogger LOG = Loggers.forClass(IndexAvailabilityController.class);
+
+    private static final String START_BUILD_INDEX_KEY_PREFIX = "startBuildIndex";
+
+    private static final String PARTITION_BUILD_INDEX_KEY_PREFIX = "partitionBuildIndex";
+
+    private final CatalogManager catalogManager;
+
+    private final MetaStorageManager metaStorageManager;
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean stopGuard = new AtomicBoolean();
+
+    /** Constructor. */
+    public IndexAvailabilityController(CatalogManager catalogManager, MetaStorageManager metaStorageManager, IndexBuilder indexBuilder) {
+        this.catalogManager = catalogManager;
+        this.metaStorageManager = metaStorageManager;
+
+        addListeners(catalogManager, metaStorageManager, indexBuilder);
+    }
+
+    @Override
+    public void close() {
+        if (!stopGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+    }
+
+    private void addListeners(CatalogService catalogService, MetaStorageManager metaStorageManager, IndexBuilder indexBuilder) {
+        catalogService.listen(CatalogEvent.INDEX_CREATE, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onIndexCreate((CreateIndexEventParameters) parameters).thenApply(unused -> false);
+        });
+
+        catalogService.listen(CatalogEvent.INDEX_DROP, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onIndexDrop((DropIndexEventParameters) parameters).thenApply(unused -> false);
+        });
+
+        catalogService.listen(CatalogEvent.INDEX_AVAILABLE, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onIndexAvailable((MakeIndexAvailableEventParameters) parameters).thenApply(unused -> false);
+        });
+
+        metaStorageManager.registerPrefixWatch(ByteArray.fromString(PARTITION_BUILD_INDEX_KEY_PREFIX), new WatchListener() {
+            @Override
+            public CompletableFuture<Void> onUpdate(WatchEvent event) {
+                return onUpdatePartitionBuildIndexKey(event).thenApply(unused -> null);
+            }
+
+            @Override
+            public void onError(Throwable e) {
+                LOG.error("Error on handle partition build index key", e);
+            }
+        });
+
+        indexBuilder.listen((indexId, tableId, partitionId) -> onIndexBuildForPartition(indexId, partitionId));
+    }
+
+    private CompletableFuture<?> onIndexCreate(CreateIndexEventParameters parameters) {
```

Comment:
Do I understand correctly that all nodes of the cluster will react on creation/removal/made-available/build-finished events, and we'll rely on MetaStorage to only allow one of the attempts to succeed?

Reply:
Yes.